docs/design.md

```diff
@@ -0,0 +1,279 @@
+        Start           uint64  // optional
+        Advertisement   Advertisement // optional
+        Entries         []Entry
+}
+```
+- `Size`: Total entries included for `AdvertisementID`.
+- `Error`: Error code (if any).
+- `Start`: Index of the first entry of the response.
+- `Advertisement`: Advertisement to which this response (and thus indexing data) belongs to.
+- `Entries`: The actual entries with the new/removed CIDs.
+
+
+```go
+type Entry struct {
+        IsRm            bool
+        Cids            []cid.Cid   // NOTE: We can probably use String instead of cid.Cid
+        Metadata        []byte
+
+}
+```
+- `IsRm`: Flags if the cids included in the entry should be added or removed from hte indexer node.G
+- `Cids`: List of CIDs included in the entry.
+- `Metadata`: Metadata to index for CIDs in the entry. If `IsRm = true` it doesn't make sense to add `Metadata` in the Entry.
+
+## Interfaces
+### Data Provider
+Interface implemented by data providers for the interaction with indexer nodes.
+
+```go
+type IngestionServer interface {
+        // Broadcast advertisement to the network
+        func Broadcast (ctx context.Context, ad Advertisement) error
+        
+        // Push an update for a single entry.
+        // This can be used to perform updates for a small number of CIDs
+        // When a full advertisement is not worth it (web3.storage case).
+        // Indexer may only accept pushes from authenticated providers.
+        func Push (ctx context.Context, indexer peer.ID, entry Entry) 
+}
+```
+
+Data providers need to include protocol handlers able to process and serve indexer node 
+ingestion requests (see [protocol](#protocol)).
+
+### Indexer Node
+Interface implemented by indexer nodes for the interaction with data providers.
+
+```go
+type IngestionClient interface{
+        // Ingest requests the ingestion of a specific range of updates
+        func Ingest (ctx context.Context, p peer.ID, firstID AdvertisementID, lastID AdvertisementID) error
+
+        // Sync with a data provider up to latest ID
+        func Sync (ctx context.Context, p peer.ID, latest AdvertisementID) error
+}
+
+```
+
+Data providers need to include protocol handlers able to process and serve indexer node 
+push Advertisements and immediate ingestions (see [protocol](#protocol)).
+
+## Protocol
+The interaction protocol between data providers and indexer nodes can be divided in two subprotocols: the advertisement protocol, and the ingestion protocol.
+
+## Advertisement Protocol
+It orchestrates the announcement of udpates from data providers to indexer nodes. Every time a data provider receives
+an update for the data they provide (e.g. new deal in a Filecoin miner), they may choose to broadcast an `Advertisement`
+to the network so indexer nodes can start indexing this data. Three kinds of updates will initially be supported:
+- `Pull` Advertisements: Data providers broadcast the `Advertisement` for the new data using a shared
```

Comment:
Pull should be split into two types:
- **Subscribe**: Indexer receives advertisement on gossip pib-sub channel
- **Poll**: Indexer periodically requests latest advertisement from provider

Reply:
Agree. Fixed!